File: Algorithms/p040_Combination_Sum_II/p040_Combination_Sum_II.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        res = []
        cur = []
        self.dfs(res,cur,candidates,target,0)
        return res

    def dfs(self, res, cur, candidates, target, i):
        last = -1
        for j in range(i,len(candidates)):
            if candidates[j]>target:
                break
            if candidates[j]==last:
                continue
            cur.append(candidates[j])
            if candidates[j]<target:
                self.dfs(res,cur,candidates,target-candidates[j],j+1)
            elif candidates[j]==target:
                res.append(cur[:])
            cur.pop()
            last = candidates[j]

File: Algorithms/p040_Combination_Sum_II/p040_Combination_Sum_II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>> res;
        vector<int> cur;
        dfs(res,cur,candidates,target,0);
        return res;
    }
    void dfs(vector<vector<int>>& res, vector<int>& cur, vector<int>& candidates, int target, int i) {
        for(int j=i,last=-1;j<candidates.size();++j){
            if(candidates[j]>target) break;
            if(candidates[j]==last) continue;
            cur.push_back(candidates[j]);
            if(candidates[j]<target){
                dfs(res,cur,candidates,target-candidates[j],j+1);
            }else if(candidates[j]==target){
                res.push_back(cur);
            }
            cur.pop_back();
            last = candidates[j];
        }
    }
};
